lib/ex_aws/sns/public_key_cache.ex

```diff
@@ -0,0 +1,59 @@
+defmodule ExAws.SNS.PublicKeyCache do
+  use GenServer
+
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, :ok, opts)
+  end
+
+  def get(certUrl) do
+      case :ets.lookup(__MODULE__, certUrl) do
+        [{certUrl, public_key}] -> {:ok, public_key}
+        [] -> GenServer.call(__MODULE__, {:get_public_key, certUrl})
+    end
+  end
+
+  ## Callbacks
+
+  def init(:ok) do
+    ets = :ets.new(__MODULE__, [:named_table, read_concurrency: true])
+    {:ok, ets}
+  end
+
+  def handle_call({:get_public_key, certUrl}, _from, ets) do
+    with {:ok, cert} <- fetch_certificate(certUrl),
+      {:ok, public_key} <- get_public_key(cert) do
+        :ets.insert(__MODULE__, {certUrl, public_key})
+        {:reply, {:ok, public_key}, ets}
+    else
+      error -> {:reply, error, ets}
+    end
+  end
+
+  defp fetch_certificate(cert_url) do
+    with {:ok, 200, _resp_headers, client} <- :hackney.get(cert_url, [], "", []),
```

Comment:
We should use `:hackney.get(cert_url, [], "", [:with_body])` because
1) it's simpler
2) Right now we only read the body on `200`. However the body doesn't go away for other status codes, and will result in basically a memory leak if not read.